File: src/internal/skillhub/service.rs
use super::agent_detector::detect_all_installed_agents;
use super::client::{load_config, save_config, Client};
use colored::Colorize;
use sha2::{Digest, Sha256};
use std::fs;
use std::io::{self, Write};
use std::path::{Path, PathBuf};

impl Client {
    pub fn find(
        &self,
        query: &str,
        category: &str,
        interactive: bool,
        cursor: &str,
    ) -> anyhow::Result<()> {
        if interactive {
            anyhow::bail!("interactive mode not implemented yet");
        }

        let display_query = if query.is_empty() { "all" } else { query };
        if !category.is_empty() {
            println!("🔍 Searching for \"{display_query}\" in category \"{category}\"\n");
        } else {
            println!("🔍 Searching for \"{display_query}\"\n");
        }

        let result = self.search_skills(query, category, cursor)?;
        if result.results.is_empty() {
            println!("No skills found.");
            return Ok(());
        }

        println!("Found {} result(s)\n", result.results.len());
        for skill in result.results {
            println!("{}", skill.display_name.cyan());
            println!("  {}", skill.description);
            println!("  {} • {}", "slug:".bright_black(), skill.slug.yellow());
            println!();
        }

        if !result.next_cursor.is_empty() {
            println!("To view more results, use: --cursor {}", result.next_cursor);
        }

        Ok(())
    }

    pub fn list(&self, category: &str, _sort: &str) -> anyhow::Result<()> {
        self.find("", category, false, "")
    }

    pub fn add(
        &self,
        slug: &str,
        version: &str,
        _global: bool,
        skip_confirm: bool,
    ) -> anyhow::Result<()> {
        println!("{} {}", "📦 Installing skill:".bold(), slug.cyan());

        let detail = self.get_skill_detail(slug)?;
        let install_version = if version.is_empty() {
            detail.latest_version.version.as_str()
        } else {
            version
        };

        println!(
            "{} {} v{}",
            "▸".bold(),
            detail.skill.display_name.bold(),
            install_version.green()
        );
        println!("  {}\n", detail.skill.description);

        println!("{}", "⬇ Downloading...".bold());
        let zip_data = self
            .download_skill(slug, install_version)
            .map_err(|err| anyhow::anyhow!("Download skill failed: {err}"))?;

        let hash = format!("{:x}", Sha256::digest(&zip_data));
        println!("  {}", format!("SHA256: {hash}").bright_black());

        let agents = detect_all_installed_agents();
        if !agents.is_empty() {
            let names: Vec<_> = agents
                .iter()
                .map(|agent| agent.display_name.as_str())
                .collect();
            println!(
                "  {} {}",
                "Detected agents:".bright_black(),
                names.join(", ").cyan()
            );
        }

        let home = home_dir();
        let global_repo = home.join(".agents").join("skills").join(slug);
        if global_repo.exists() {
            if !skip_confirm {
                print!("Skill {} already exists. Overwrite? [y/N] ", slug.cyan());
                io::stdout().flush()?;
                let mut answer = String::new();
                io::stdin().read_line(&mut answer)?;
                if !answer.trim().eq_ignore_ascii_case("y") {
                    println!("{}", "Installation cancelled.".yellow());
                    return Ok(());
                }
            }
            let _ = fs::remove_dir_all(&global_repo);
        }

        println!("{}", "📂 Extracting to global repository...".bold());
        extract_zip(&zip_data, &global_repo)
            .map_err(|err| anyhow::anyhow!("Extract failed: {err}"))?;

        if cfg!(windows) {
            println!("{}", "📋 Copying to agent directories...".bold());
        } else {
            println!("{}", "🔗 Creating symlinks...".bold());
        }

        let mut linked_count = 0usize;
        for agent in agents {
            let link_path = agent.global_skills_dir.join(slug);
            if link_path == global_repo {
                continue;
            }
            if let Err(err) = fs::create_dir_all(&agent.global_skills_dir) {
                println!(
                    "  {} Failed to create directory for {}: {}",
                    "⚠".yellow(),
                    agent.display_name,
                    err
                );
                continue;
            }
            let _ = remove_path(&link_path);
            if let Err(err) = link_or_copy_dir(&global_repo, &link_path) {
                println!(
                    "  {} Failed to link to {}: {}",
                    "⚠".yellow(),
                    agent.display_name,
                    err
                );
            } else {
                println!(
                    "  {} {} -> {}",
                    "✓".green(),
                    agent.display_name.cyan(),
                    link_path.display().to_string().bright_black()
                );
                linked_count += 1;
            }
        }

        println!();
        println!("{}", "✅ Installation complete!".bold().green());
        println!(
            "  {} {}",
            "Global repository:".bright_black(),
            global_repo.display().to_string().cyan()
        );
        if linked_count > 0 {
            let label = if cfg!(windows) {
                "Copies:"
            } else {
                "Symlinks:"
            };
            println!(
                "  {} Linked to {} agent(s)",
                label.bright_black(),
                linked_count
            );
        }
        println!();

        println!("{}", "Usage:".bold());
        println!("  The skill is now available in all your agent sessions.");
        println!("  Restart your agent if needed to load the new skill.");

        Ok(())
    }

    pub fn config(&self, set_url: &str, show: bool) -> anyhow::Result<()> {
        let mut config = load_config()?;
        if !set_url.is_empty() {
            config.api_base_url = set_url.to_string();
            save_config(&config)?;
            println!("{} API URL updated to: {}", "✓".green(), set_url.cyan());
            return Ok(());
        }

        if show {
            println!("{}", "Current Configuration:".bold());
            println!(
                "  {} {}",
                "API URL:".bright_black(),
                config.api_base_url.cyan()
            );
            println!(
                "  {} {}",
                "Config file:".bright_black(),
                super::client::config_file_path()
                    .display()
                    .to_string()
                    .cyan()
            );
            return Ok(());
        }

        anyhow::bail!("use --set-url or --show")
    }
}

fn home_dir() -> PathBuf {
    std::env::var_os("HOME")
        .map(PathBuf::from)
        .or_else(dirs::home_dir)
        .unwrap_or_else(|| PathBuf::from("."))
}

fn remove_path(path: &Path) -> anyhow::Result<()> {
    if !path.exists() {
        return Ok(());
    }
    let metadata = fs::symlink_metadata(path)?;
    if metadata.is_dir() && !metadata.file_type().is_symlink() {
        fs::remove_dir_all(path)?;
    } else {
        fs::remove_file(path)?;
    }
    Ok(())
}

fn link_or_copy_dir(src: &Path, dst: &Path) -> anyhow::Result<()> {
    if cfg!(windows) {
        copy_dir(src, dst)
    } else {
        #[cfg(unix)]
        {
            std::os::unix::fs::symlink(src, dst)?;
            Ok(())
        }
        #[cfg(not(unix))]
        {
            copy_dir(src, dst)
        }
    }
}

fn copy_dir(src: &Path, dst: &Path) -> anyhow::Result<()> {
    let metadata = fs::metadata(src)?;
    fs::create_dir_all(dst)?;
    fs::set_permissions(dst, metadata.permissions())?;
    for entry in fs::read_dir(src)? {
        let entry = entry?;
        let src_path = entry.path();
        let dst_path = dst.join(entry.file_name());
        if entry.file_type()?.is_dir() {
            copy_dir(&src_path, &dst_path)?;
        } else {
            copy_file(&src_path, &dst_path)?;
        }
    }
    Ok(())
}

fn copy_file(src: &Path, dst: &Path) -> anyhow::Result<()> {
    if let Some(parent) = dst.parent() {
        fs::create_dir_all(parent)?;
    }
    fs::copy(src, dst)?;
    fs::set_permissions(dst, fs::metadata(src)?.permissions())?;
    Ok(())
}

fn extract_zip(zip_data: &[u8], target_dir: &Path) -> anyhow::Result<()> {
    let reader = std::io::Cursor::new(zip_data);
    let mut archive = zip::ZipArchive::new(reader)?;

    let mut top_level_dirs = std::collections::BTreeSet::new();
    for i in 0..archive.len() {
        let file = archive.by_index(i)?;
        if let Some(first) = file.name().split('/').find(|part| !part.is_empty()) {
            top_level_dirs.insert(first.to_string());
        }
    }
    let strip_prefix = if top_level_dirs.len() == 1 {
        top_level_dirs.iter().next().map(|dir| format!("{dir}/"))
    } else {
        None
    };

    for i in 0..archive.len() {
        let mut file = archive.by_index(i)?;
        let mut file_name = file.name().to_string();
        if let Some(prefix) = &strip_prefix {
            file_name = file_name
                .strip_prefix(prefix)
                .unwrap_or(&file_name)
                .to_string();
        }
        if file_name.is_empty() {
            continue;
        }
        let target_path = target_dir.join(&file_name);
        if !target_path.starts_with(target_dir) {
            anyhow::bail!("zip entry escapes target directory: {file_name}");
        }

        if file.is_dir() {
            fs::create_dir_all(&target_path)?;
            continue;
        }

        if let Some(parent) = target_path.parent() {
            fs::create_dir_all(parent)?;
        }
        let mut out = fs::File::create(&target_path)?;
        io::copy(&mut file, &mut out)?;
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::internal::test_support::{TestResponse, TestServer};
    use serial_test::serial;
    use std::env;
    use tempfile::TempDir;
    use zip::write::FileOptions;

    fn zip_with_top_level() -> Vec<u8> {
        let cursor = std::io::Cursor::new(Vec::new());
        let mut zip = zip::ZipWriter::new(cursor);
        zip.add_directory("skill/", FileOptions::default()).unwrap();
        zip.start_file("skill/SKILL.md", FileOptions::default())
            .unwrap();
        zip.write_all(b"# Skill").unwrap();
        zip.finish().unwrap().into_inner()
    }

    #[test]
    fn extract_zip_strips_single_top_level_directory_and_copy_file() {
        let target = TempDir::new().unwrap();
        extract_zip(&zip_with_top_level(), target.path()).unwrap();
        assert_eq!(
            fs::read_to_string(target.path().join("SKILL.md")).unwrap(),
            "# Skill"
        );

        let copied = target.path().join("copy.md");
        copy_file(&target.path().join("SKILL.md"), &copied).unwrap();
        assert_eq!(fs::read_to_string(copied).unwrap(), "# Skill");
    }

    #[test]
    fn copy_dir_and_link_or_copy_dir_work_for_directories() {
        let root = TempDir::new().unwrap();
        let src = root.path().join("src");
        let dst = root.path().join("dst");
        fs::create_dir_all(&src).unwrap();
        fs::write(src.join("file.txt"), "ok").unwrap();
        copy_dir(&src, &dst).unwrap();
        assert_eq!(fs::read_to_string(dst.join("file.txt")).unwrap(), "ok");

        let linked = root.path().join("linked");
        link_or_copy_dir(&src, &linked).unwrap();
        assert!(linked.exists());
        remove_path(&linked).unwrap();
        assert!(!linked.exists());
    }

    #[test]
    #[serial]
    fn find_list_config_and_add_happy_path() {
        let home = TempDir::new().unwrap();
        env::set_var("HOME", home.path());
        fs::create_dir_all(home.path().join(".codex")).unwrap();
        let zip = zip_with_top_level();
        let server = TestServer::new(move |req| match req.path.as_str() {
            "/search?q=&limit=20" => TestResponse::json(
                200,
                r#"{"results":[{"slug":"cat","displayName":"Cat","description":"desc"}]}"#,
            ),
            "/skills/cat" => TestResponse::json(
                200,
                r#"{"skill":{"slug":"cat","displayName":"Cat","description":"desc"},"latestVersion":{"version":"1.0.0"}}"#,
            ),
            "/skills/cat/download?version=1.0.0" => {
                TestResponse::bytes(200, zip.clone(), "application/zip")
            }
            other => panic!("unexpected path {other}"),
        });
        env::set_var("SEACLOUD_SKILLHUB_URL", server.url());
        let client = Client::new();
        client.find("", "", false, "").unwrap();
        client.list("", "").unwrap();
        client
            .config("https://skillhub.example.com/api", false)
            .unwrap();
        client.config("", true).unwrap();
        client.add("cat", "", true, true).unwrap();
        assert!(home.path().join(".agents/skills/cat/SKILL.md").exists());
    }

    #[test]
    fn interactive_and_empty_config_errors_are_reported() {
        let client = Client::new();
        assert!(client
            .find("", "", true, "")
            .unwrap_err()
            .to_string()
            .contains("interactive"));
        assert!(client
            .config("", false)
            .unwrap_err()
            .to_string()
            .contains("use --set-url"));
    }
}

File: internal/skillhub/service.go
package skillhub

import (
	"archive/zip"
	"bytes"
	"crypto/sha256"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"runtime"
	"strings"

	"github.com/fatih/color"
)

func (c *Client) Add(slug, version string, global, skipConfirm bool) error {
	fmt.Printf("%s %s\n", color.New(color.Bold).Sprint("📦 Installing skill:"), color.CyanString(slug))

	detail, err := c.GetSkillDetail(slug)
	if err != nil {
		return err
	}

	installVersion := version
	if installVersion == "" {
		installVersion = detail.LatestVersion.Version
	}

	fmt.Printf("%s %s v%s\n",
		color.New(color.Bold).Sprint("▸"),
		color.New(color.Bold).Sprint(detail.Skill.DisplayName),
		color.GreenString(installVersion),
	)
	fmt.Printf("  %s\n\n", detail.Skill.Description)

	fmt.Println(color.New(color.Bold).Sprint("⬇ Downloading..."))
	zipData, err := c.DownloadSkill(slug, installVersion)
	if err != nil {
		return fmt.Errorf("Download skill failed: %w", err)
	}

	hash := fmt.Sprintf("%x", sha256.Sum256(zipData))
	fmt.Printf("  %s\n", color.New(color.FgHiBlack).Sprintf("SHA256: %s", hash))

	// Detect all installed agents
	agents := DetectAllInstalledAgents()
	if len(agents) > 0 {
		fmt.Printf("  %s ",
			color.New(color.FgHiBlack).Sprint("Detected agents:"),
		)
		var names []string
		for _, a := range agents {
			names = append(names, a.DisplayName)
		}
		fmt.Printf("%s\n", color.CyanString(strings.Join(names, ", ")))
	}

	// Install to global repository
	home, _ := os.UserHomeDir()
	globalRepo := filepath.Join(home, ".agents", "skills", slug)

	if _, err := os.Stat(globalRepo); err == nil {
		if !skipConfirm {
			fmt.Printf("Skill %s already exists. Overwrite? [y/N] ", color.CyanString(slug))
			var answer string
			fmt.Scanln(&answer)
			if strings.ToLower(answer) != "y" {
				fmt.Println(color.YellowString("Installation cancelled."))
				return nil
			}
		}
		_ = os.RemoveAll(globalRepo)
	}

	fmt.Println(color.New(color.Bold).Sprint("📂 Extracting to global repository..."))
	if err := extractZip(zipData, globalRepo); err != nil {
		return fmt.Errorf("Extract failed: %w", err)
	}

	// Create symlinks (Unix/macOS) or copies (Windows) for all detected agents
	if runtime.GOOS == "windows" {
		fmt.Println(color.New(color.Bold).Sprint("📋 Copying to agent directories..."))
	} else {
		fmt.Println(color.New(color.Bold).Sprint("🔗 Creating symlinks..."))
	}

	linkedCount := 0
	for _, agent := range agents {
		linkPath := filepath.Join(agent.GlobalSkillsDir, slug)

		// Skip if the link path is the same as global repo (e.g., Cline uses ~/.agents/skills)
		if linkPath == globalRepo {
			continue
		}

		// Ensure the skills directory exists
		if err := os.MkdirAll(agent.GlobalSkillsDir, 0755); err != nil {
			fmt.Printf("  %s Failed to create directory for %s: %v\n",
				color.YellowString("⚠"),
				agent.DisplayName,
				err,
			)
			continue
		}

		// Remove old symlink/directory if exists
		_ = os.RemoveAll(linkPath)

		// Create symlink (Unix/macOS) or copy directory (Windows)
		if err := linkOrCopyDir(globalRepo, linkPath); err != nil {
			fmt.Printf("  %s Failed to link to %s: %v\n",
				color.YellowString("⚠"),
				agent.DisplayName,
				err,
			)
		} else {
			fmt.Printf("  %s %s -> %s\n",
				color.GreenString("✓"),
				color.CyanString(agent.DisplayName),
				color.New(color.FgHiBlack).Sprint(linkPath),
			)
			linkedCount++
		}
	}

	fmt.Println()
	fmt.Println(color.New(color.Bold, color.FgGreen).Sprint("✅ Installation complete!"))
	fmt.Printf("  %s %s\n",
		color.New(color.FgHiBlack).Sprint("Global repository:"),
		color.CyanString(globalRepo),
	)
	if linkedCount > 0 {
		linkLabel := "Symlinks:"
		if runtime.GOOS == "windows" {
			linkLabel = "Copies:"
		}
		fmt.Printf("  %s Linked to %d agent(s)\n",
			color.New(color.FgHiBlack).Sprint(linkLabel),
			linkedCount,
		)
	}
	fmt.Println()

	fmt.Println(color.New(color.Bold).Sprint("Usage:"))
	fmt.Println("  The skill is now available in all your agent sessions.")
	fmt.Println("  Restart your agent if needed to load the new skill.")

	return nil
}

func extractZip(zipData []byte, targetDir string) error {
	reader, err := zip.NewReader(bytes.NewReader(zipData), int64(len(zipData)))
	if err != nil {
		return err
	}

	topLevelDirs := make(map[string]bool)
	for _, file := range reader.File {
		parts := strings.Split(file.Name, "/")
		if len(parts) > 0 && parts[0] != "" {
			topLevelDirs[parts[0]] = true
		}
	}

	var stripPrefix string
	if len(topLevelDirs) == 1 {
		for dir := range topLevelDirs {
			stripPrefix = dir + "/"
		}
	}

	for _, file := range reader.File {
		fileName := file.Name
		if stripPrefix != "" {
			fileName = strings.TrimPrefix(fileName, stripPrefix)
		}
		if fileName == "" {
			continue
		}

		targetPath := filepath.Join(targetDir, fileName)
		if file.FileInfo().IsDir() {
			_ = os.MkdirAll(targetPath, 0755)
			continue
		}

		if err := os.MkdirAll(filepath.Dir(targetPath), 0755); err != nil {
			return err
		}

		rc, err := file.Open()
		if err != nil {
			return err
		}

		outFile, err := os.Create(targetPath)
		if err != nil {
			rc.Close()
			return err
		}

		_, err = io.Copy(outFile, rc)
		rc.Close()
		outFile.Close()
		if err != nil {
			return err
		}
	}

	return nil
}
